toolchain/base/timings.h

```diff
@@ -14,9 +14,33 @@ namespace Carbon {
 // Helps track timings for a compile.
 class Timings {
  public:
+  // Records a timing for a scope, if the timings object is valid.
+  class ScopedTiming {
+   public:
+    explicit ScopedTiming(std::optional<Timings>* timings,
+                          llvm::StringRef label)
+        : timings(timings),
+          label(label),
+          start(*timings ? std::chrono::steady_clock::now()
+                         : std::chrono::steady_clock::time_point::min()) {}
+
+    ~ScopedTiming() {
+      if (*timings) {
+        (*timings)->Add(label, std::chrono::steady_clock::now() - start);
+      }
+    }
+
+   private:
+    std::optional<Timings>* timings;
+    llvm::StringRef label;
```

Comment:
WDYT about using a `std::string` here and moving it into the `Add` call (with the old API)?

Mostly it'd make it safe to use this with a string built with any formatting, and it shouldn't change how many strings we end up needing.

Reply:
That would allocate the string regardless of whether timings were on. Is that desirable?

Reply:
Note, might want a slightly different structure if this is used with formatting, since we'd probably want to avoid creation of the formatted string when not storing timings.